qutebrowser/misc/sql.py

```diff
@@ -168,75 +169,160 @@ def raise_sqlite_error(msg, error):
+
+    def __init__(self, path: str):
+        if QSqlDatabase.database(path).isValid():
+            raise BugError(f'A connection to the database at "{path}" already exists')
+
+        self._path = path
+        database = QSqlDatabase.addDatabase('QSQLITE', path)
+        if not database.isValid():
+            raise KnownError('Failed to add database. Are sqlite and Qt sqlite '
+                             'support installed?')
+        database.setDatabaseName(path)
+        if not database.open():
+            error = database.lastError()
+            msg = f"Failed to open sqlite database at {path}: {error.text()}"
+            raise_sqlite_error(msg, error)
+
+        version_int = self.query('pragma user_version').run().value()
+        self._user_version = UserVersion.from_int(version_int)
+
+        if self._user_version.major > self._USER_VERSION.major:
+            raise KnownError(
+                "Database is too new for this qutebrowser version (database version "
+                f"{self._user_version}, but {self._USER_VERSION.major}.x is supported)")
+
+        if self.user_version_changed():
+            log.sql.debug(f"Migrating from version {self._user_version} "
+                          f"to {self._USER_VERSION}")
+            # Note we're *not* updating _user_version here. We still want
+            # user_version_changed() to return True, as other modules (such as
+            # history.py) use it to create the initial table structure.
+            self.query(f'PRAGMA user_version = {_USER_VERSION.to_int()}').run()
+
+            # Enable write-ahead-logging and reduce disk write frequency
+            # see https://sqlite.org/pragma.html and issues #2930 and #3507
+            #
+            # We might already have done this (without a migration) in earlier versions,
+            # but as those are idempotent, let's make sure we run them once again.
+            self.query("PRAGMA journal_mode=WAL").run()
+            self.query("PRAGMA synchronous=NORMAL").run()
+
+    def qSqlDatabase(self):
+        database = QSqlDatabase.database(self._path, open=True)
+        if not database.isValid():
+            raise BugError('Failed to get connection. Did you close() this Database '
+                           'instance?')
+        return database
+
+    def query(self, querystr, forward_only=True):
+        return Query(self, querystr, forward_only)
+
+    def table(self, name, fields, constraints=None, parent=None):
+        # Still need to link the SqlTable object to this Database instance
+        return SqlTable(self, name, fields, constraints, parent)
+
+    def user_version_changed(self):
+        """Whether the version stored in the database differs from the current one."""
+        return self._user_version != self._USER_VERSION
+
+    def close(self):
+        """Close the SQL connection."""
+        database = self.qSqlDatabase()
+        database.close()
+        del database
```

Comment:
The next statement will raise a warning if I don't do this.

Reply:
What warning exactly?

Depending on what the warning is, using [`sip.delete`](https://www.riverbankcomputing.com/static/Docs/PyQt5/api/sip/sip-module.html#delete) to make sure the underling C++ object is deleted might be cleaner.

Reply:
This was the warning: `19:26:46 WARNING  qt         Unknown module:none:0 QSqlDatabasePrivate::removeDatabase: connection ':memory:' is still in use, all queries will cease to work.`

`sip.delete` also works. Thanks.